---
title: "RNA-seq analysis of WT and cbf11 mutant in YES and YES+ammonium"
author: "Martin Převorovský, prevorov@natur.cuni.cz, http://www.GenoMik.cz"
output: html_document
---

```{r setup, message=FALSE}
library(rtracklayer)
library(GenomicAlignments)
library(DESeq2)

knitr::opts_chunk$set(echo = TRUE)

genome_dir <- './genome/'
bam_dir <- './BAM_RNA-seq/'
coverage_dir <- './coverage_RNA-seq/'
QC_dir <- './QC_RNA-seq/'
QC_trim_dir <- './QC_trim_RNA-seq/'
image_dir <- './images/'
deg_dir <- './DEG_RNA-seq/'
genome_gff <- paste0(genome_dir, 'Schizosaccharomyces_pombe_all_chromosomes.gff3')
```

Variables for shell scripts are stored in the file "shell_variables_RNA-seq.sh" and can be modified as needed:

```{bash list_variables}
cat shell_variables_RNA-seq.sh
```


# Getting the raw RNA-seq read datasets

THIS SECTION WILL ONLY BECOME RELEVANT ONCE THE SEQUENCING DATA ARE PUBLIC in ARRAYEXPRESS

All data and metadata can be obtained from ArrayExpress at https://www.ebi.ac.uk/arrayexpress/experiments/XXXXXXXX/.

```{bash, eval=FALSE, message=FALSE}
source shell_variables_RNA-seq.sh

wget -N --no-verbose https://www.ebi.ac.uk/arrayexpress/files/XXXXXXXX/XXXXXXXX.sdrf.txt
mapfile -t ftp_links < <(tail -n +2 XXXXXXXX.sdrf.txt | cut -f30)
mapfile -t ftp_names < <(tail -n +2 XXXXXXXX.sdrf.txt | cut -f28)

mkdir -p "${fastq_dir}"
for i in ${!ftp_links[@]};
do
  wget --no-verbose -O "${fastq_dir}${ftp_names[${i}]}" "${ftp_links[${i}]}"
done
```


# *Schizosaccharomyces pombe* genome sequence and annotation

* genome was downloaded from ftp://ftp.pombase.org/pombe/genome_sequence_and_features/genome_sequence/ (downloaded `r date()`)

* genome annotation (GFF): ftp://ftp.pombase.org/pombe/genome_sequence_and_features/gff3/ (downloaded `r date()`)

```{bash}
bedtools --version
```

```{bash, message=FALSE, eval=TRUE}
source shell_variables_RNA-seq.sh

mkdir -p "${genome_dir}"
wget -N --no-verbose --directory-prefix="${genome_dir}"  ftp://ftp.pombase.org/pombe/genome_sequence_and_features/genome_sequence/Schizosaccharomyces_pombe_chromosome_I.fa.gz
wget -N --no-verbose --directory-prefix="${genome_dir}"  ftp://ftp.pombase.org/pombe/genome_sequence_and_features/genome_sequence/Schizosaccharomyces_pombe_chromosome_II.fa.gz
wget -N --no-verbose --directory-prefix="${genome_dir}"  ftp://ftp.pombase.org/pombe/genome_sequence_and_features/genome_sequence/Schizosaccharomyces_pombe_chromosome_III.fa.gz
wget -N --no-verbose --directory-prefix="${genome_dir}" ftp://ftp.pombase.org/pombe/genome_sequence_and_features/genome_sequence/Schizosaccharomyces_pombe_mitochondrial_chromosome.fa.gz
zcat `ls "${genome_dir}"* | grep fa.gz$` > "${genome}"

wget -N --no-verbose --directory-prefix="${genome_dir}"  ftp://ftp.pombase.org/pombe/genome_sequence_and_features/gff3/Schizosaccharomyces_pombe_all_chromosomes.gff3.gz
gzip -fd "${genome_dir}Schizosaccharomyces_pombe_all_chromosomes.gff3.gz"

samtools faidx "${genome}"
cut -f1,2 "${genome}.fai" > "${genome_dir}chrom.sizes"
```


# QC of raw sequence reads

Technical quality of library construction and sequencing run was tested using fastQC

```{bash}
fastqc --version
multiqc --version
```

```{bash, eval=TRUE}
source shell_variables_RNA-seq.sh

mkdir -p "${QC_dir}"
mapfile -t fastq_files < <(ls -1 "${fastq_dir}" | grep "${fastq_file_ext}")
for i in ${fastq_files[@]};
do
  fastqc --quiet --outdir "${QC_dir}" --threads "${CPU}" "${fastq_dir}${i}"
done

multiqc -f -m fastqc -i RNA-seq_ammonium -o "${QC_dir}" "${QC_dir}"
```

Results of QC:

* 51 nt PE reads

* 36-53.1M read pairs per sample

* no adapter contamination detected here

* all samples give error/warning on high read duplication levels (normal and OK for RNA-seq from a small genome)

* all 24 FASTQ files give error on per base sequence content (positions 1-10)

* GC content is OK for all samples

* one sample gives warning on overrepresented sequences (polyG)

* read quality scores are good for all positions, all samples. Trimming is not needed


# Read mapping into *S. pombe* genome

```{bash}
hisat2 --version
samtools --version
```

```{bash, eval=TRUE}
source shell_variables_RNA-seq.sh

# build HISAT2 index
hisat2-build --quiet -p "${CPU}" "${genome}" "${genome}"

# map
mkdir -p "${bam_dir}"
fastq_files_1=( $(ls -1 "${fastq_dir}" | grep "_R1_.*${fastq_file_ext}") )
fastq_files_2=( $(ls -1 "${fastq_dir}" | grep "_R2_.*${fastq_file_ext}") )
for i in ${!fastq_files_1[@]};
do
	infile1="${fastq_dir}${fastq_files_1[${i}]}"
	infile2="${fastq_dir}${fastq_files_2[${i}]}"
	outfile="${bam_dir}${fastq_files_1[${i}]}.bam"
  echo "${outfile}"
	hisat2 -x "${genome}" -1 "${infile1}" -2 "${infile2}" --threads "${CPU}" --max-intronlen 10000 -k 15 --summary-file "${outfile}.log" | samtools view -b -F 256 --threads "${CPU}" - | samtools sort --threads "${CPU}" -o "${outfile}" - 
	samtools index "${outfile}"
	samtools view -c --threads "${CPU}" "${outfile}"
done
```

Notes

* since signal is expected at repeat regions, HISAT was set to search for up to 15 matching positions (this allows even for testing all Tf2 and dh/dg repeats in the genome), then only the best (marked as primary by HISAT) alignment was kept


# QC of mapped samples

WARNING - completion of this step can take A LOT of time (days...)!

```{bash}
deeptools --version
```

```{bash, eval=FALSE}
source shell_variables_RNA-seq.sh

mkdir -p "${image_dir}"
bam_files=( $(ls -1 -d "${bam_dir}"* | grep "\.bam$") )
multiBamSummary bins --binSize "${bin_size}" -p "${CPU}" --ignoreDuplicates -o "${bam_dir}${bam_summary_file}" --bamfiles "${bam_files[@]}"
plotPCA -in "${bam_dir}${bam_summary_file}" -o "${image_dir}RNA-seq_plotPCA.pdf"
plotCorrelation -in "${bam_dir}${bam_summary_file}" -o "${image_dir}RNA-seq_plotCorrelation.pdf" --corMethod pearson --whatToPlot heatmap --removeOutliers
```

* Samples cluster very well by genotype; there are some batch effects affecting clustering according to growth media type in WT.


# Calculate genome coverage tracks

```{bash, eval=TRUE}
source shell_variables_RNA-seq.sh

mkdir -p "${coverage_dir}"
mapfile -t bam_files < <(ls -1 "${bam_dir}" | grep "\.bam$")
for i in ${!bam_files[@]};
do
	input_file="${bam_dir}${bam_files[${i}]}"
	bamCoverage --binSize "${bin_size}" --normalizeUsing CPM -p "${CPU}" --ignoreDuplicates --bam "${input_file}" -o "${coverage_dir}${bam_files[${i}]}.bw"
done

# average coverage tracks across biological replicates
mapfile -t bw_files < <(ls -1 "${coverage_dir}" | grep "\.bam\.bw$")
samples=("WT_YES" "WT_YES+AMM" "cbf11_YES" "cbf11_YES+AMM")
for i in {0..3};
do
  wiggletools mean "${coverage_dir}${bw_files[${i}]}" "${coverage_dir}${bw_files[${i}+4]}" "${coverage_dir}${bw_files[${i}+8]}" | wigToBigWig -itemsPerSlot="${bin_size}" stdin "${genome_dir}chrom.sizes" "${coverage_dir}${samples[${i}]}.mean.bw"
done
```

Notes

* Checked the genotypes visually in IGV → OK.


# Identify differentially expressed genes

## Generate gene-level counts tables.

* Raw count data are needed for DEG calling in `DESeq2`.

* TPM-normalized (Transcripts Per Million) count data are suitable for general comparisons of the samples.

```{r, eval=TRUE}
gff <- import.gff3(genome_gff)
gff.genes <- gff[gff$type == 'gene', ]
bam_files <- list.files(bam_dir, full.names = TRUE, pattern = '\\.bam$')
bam <- BamFileList(bam_files, yieldSize = 1000000)

sample_info <- read.delim('sampleInfo', stringsAsFactors = FALSE, header = TRUE, sep = '\t')
sample_info[, 'cultivation_run'] <- factor(sample_info[, 'cultivation_run'])
sample_info[, 'genotype'] <- factor(sample_info[, 'genotype'])
sample_info[, 'medium'] <- factor(sample_info[, 'medium'])
names(bam) <- sample_info$sample_ID

se <- summarizeOverlaps(gff.genes,
                        bam,
                        mode = 'Union',
                        singleEnd = FALSE,
                        ignore.strand = FALSE,
                        fragments = FALSE,
                        preprocess.reads = invertStrand)
rownames(se) <- gff.genes$ID
colData(se) <- cbind(colData(se), sample_info)
#save(se, file = 'se.rda')
#load(file = 'se.rda')

counts <- assay(se)

# normalize to get TPM values (transcripts per million)
gene_widths <- width(gff.genes)
counts.norm <- counts
for (i in 1:nrow(counts.norm)){
  counts.norm[i, ] <- counts.norm[i, ] / gene_widths[i]
}
count_sums <- colSums(counts.norm) / 1000000
for (i in 1:length(count_sums)){
  counts.norm[, i] <- counts.norm[, i] / count_sums[i]
}

counts <- cbind(gff.genes$Name, counts)
counts.norm <- cbind(gff.genes$Name, counts.norm)
colnames(counts)[1] <- 'gene_name'
colnames(counts.norm)[1] <- 'gene_name'
write.csv(counts, file = paste0(coverage_dir, 'counts.csv'), quote = FALSE)
write.csv(counts.norm, file = paste0(coverage_dir, 'counts.TPM-normalized.csv'), quote = FALSE)
rm(gene_widths, count_sums)
```

## DEG calling

```{r, eval=TRUE}
# Benjamini-Hochberg p value adjustment (FDR)
padj_threshold <- 0.05

dir.create(deg_dir)

# cbf11-YES vs WT-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + genotype)
dds <- dds[, dds$medium =='YES'] # remove YES+ammonium samples ('YES_AMM')
dds$medium <- droplevels(dds$medium) # remove factor levels without samples
dds$genotype <- relevel(dds$genotype, ref = 'WT') # change order of factor levels to get cbf11/WT fold change (= define the reference)
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_cbf11-YES_vs_WT-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_cbf11-YES_vs_WT-YES.SIG.csv'), quote = FALSE)

# cbf11-YES+AMM vs WT-YES+AMM
dds <- DESeqDataSet(se, design = ~ cultivation_run + genotype)
dds <- dds[, dds$medium =='YES_AMM']
dds$medium <- droplevels(dds$medium)
dds$genotype <- relevel(dds$genotype, ref = 'WT')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_cbf11-YES+AMM_vs_WT-YES+AMM.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_cbf11-YES+AMM_vs_WT-YES+AMM.SIG.csv'), quote = FALSE)

# WT-YES+AMM vs WT-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + medium)
dds <- dds[, dds$genotype =='WT']
dds$genotype <- droplevels(dds$genotype)
dds$medium <- relevel(dds$medium, ref = 'YES')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_WT-YES+AMM_vs_WT-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_WT-YES+AMM_vs_WT-YES.SIG.csv'), quote = FALSE)

# cbf11-YES+AMM vs cbf11-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + medium)
dds <- dds[, dds$genotype =='cbf11']
dds$genotype <- droplevels(dds$genotype)
dds$medium <- relevel(dds$medium, ref = 'YES')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_cbf11-YES+AMM_vs_cbf11-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_cbf11-YES+AMM_vs_cbf11-YES.SIG.csv'), quote = FALSE)
```

***

```{r}
sessionInfo()
```
